Refactoring mattermost package, to suit for distribution

phoinixgrr opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
We are introducing the following essential changes for `matermost` package, to better suit for distribution: 

- Change binary destination to `/usr/bin/` instead of `/usr/local/bin/` , to align with best practises. 
 `/usr/bin` is for distribution-managed normal user programs.
  `/usr/local/bin` is for normal user programs not managed by the distribution package manager, e.g. locally compiled packages. 
- Renaming config file from `config.json` to `config.defaults.json` to keep in case of upgrade/purge remove 
- Introduce logic, to define if this is a first time installation, or an upgrade 
- Create mattermost OS user/grp and define relevant permissions 
- Remove mattermost OS user/grp in case of removal
- Perform file cleanup(old mm version) in case of upgrade 
- Introduce env variable:`MM_INSTALL_TYPE` under runtime
- Remove postgresql dependency for `mattermost.service`


#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
**Ticket:** https://mattermost.atlassian.net/browse/CLD-4311

spirosoik commented on Makefile:
```diff
@@ -48,11 +48,15 @@ mattermost_$(version)-$(revision).deb:
+	   ${build_path}/opt/mattermost/config/config.defaults.json
+
 	@echo "Creating binary links"
-	mkdir -p ${build_path}/usr/local/bin
```
won't this change affect omnibus? I dunno that's why I am asking

phoinixgrr replied:
Nope. Omnibus is tested and works as expected. 
mattermost binary symbolic link is still available on the OS

spirosoik commented on mattermost/DEBIAN/postinst:
```diff
@@ -1,7 +1,36 @@
 case "$1" in
     configure)
+
+    [ -z "$MATTERMOST_USER" ] && MATTERMOST_USER="mattermost"
```
I think you are trying to do this? 
`MATTERMOST_USER=${MATTERMOST_USER:-mattermost}`

same for group

phoinixgrr replied:
We are trying to perform the following: 
If `MATTERMOST_USER` is null (not set) we set ot to `mattermost` 
Same for `MATTERMOST_GROUP`

spirosoik replied:
yeah this is my intent too with what I shared. you can give a try

phoinixgrr replied:
Tested it, seems to work as expected. 

What do we gain with the proposal? 
Wouldn't you agree that it is a more readable form? :
`[ -z "$MATTERMOST_USER" ] && MATTERMOST_USER="mattermost"` 
vs 
`MATTERMOST_USER=${MATTERMOST_USER:-mattermost}` 

The latter requires some bash insights.  

Would vote in favor of readability against saving a couple of chars. 

Let me know what do you think

spirosoik replied:
I like the second opinion but 0/5 not strong opinion. minor thing to change

stafot replied:
Alos prefer the 2nd one, but also 0/5

spirosoik commented on mattermost/DEBIAN/postinst:
```diff
@@ -1,7 +1,36 @@
 #!/usr/bin/env bash
 
+set -e
+IS_UPGRADE=false
```
nit: you can do something like

```
TRUE_REG='^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$'
FALSE_REG='^([fF][aA][lL][sS][eE]|[nN]|[nN][oO]|0)$'
```

and then do:
```
if [[ $IS_UPGRADE =~ $TRUE_REG ]]; then
```

Also is a good practice to add some debug flag and strict scripting to help us debug when is needed:
```
IS_DEBUG=${IS_DEBUG:-false}
if [[ $IS_DEBUG =~ $TRUE_REG ]]; then
    set -o xtrace
fi

IS_STRICT=${IS_STRICT:-false}
if [[ $IS_STRICT =~ $TRUE_REG ]]; then
    set -o errexit
    set -o nounset
    set -o pipefail
fi
```

phoinixgrr replied:
These are getting shipped with the mattermost debian packages. Are we sure we want to introduce debugging flags to our released packages? 

Example 
https://github.com/grafana/grafana/blob/main/packaging/deb/control/postinst

spirosoik replied:
Yeah I got it but I believe this will be really helpful if a customer mentions it doesn't work and somehow be able to debug and get some details back in which line etc. failed. In an on-prem environment will be super hard to debug. Thoughts?

phoinixgrr replied:
Valid point. 👍 
Will push a Commit to support debugging

phoinixgrr replied:
Done with: 


```bash
> sudo IS_DEBUG=true dpkg -i mattermost_7.3.1-0.deb
Selecting previously unselected package mattermost.
(Reading database ... 62439 files and directories currently installed.)
Preparing to unpack mattermost_7.3.1-0.deb ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
Unpacking mattermost (7.3.1-0) ...
Setting up mattermost (7.3.1-0) ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ '[' -z '' ']'
+ MATTERMOST_USER=mattermost
+ '[' -z '' ']'
+ MATTERMOST_GROUP=mattermost
+ getent group mattermost
+ addgroup --system mattermost --quiet
+ id mattermost
+ adduser --system --home /opt/mattermost --no-create-home --ingroup mattermost --disabled-password --shell /bin/false mattermost
+ chown -R mattermost:mattermost /opt/mattermost
+ '[' '' '!=' '' ']'
+ [[ ! false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ systemctl daemon-reload
+ systemctl enable mattermost
> sudo IS_DEBUG=true dpkg -i mattermost_7.4.0-0.deb
(Reading database ... 66288 files and directories currently installed.)
Preparing to unpack mattermost_7.4.0-0.deb ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ systemctl stop mattermost
+ '[' 7.3.1-0 '!=' '' ']'
+ IS_UPGRADE=true
+ [[ true =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ find /opt/mattermost/ /opt/mattermost/client/ -mindepth 1 -maxdepth 1 '!' '(' -type d '(' -path /opt/mattermost/client -o -path /opt/mattermost/client/plugins -o -path /opt/mattermost/config -o -path /opt/mattermost/logs -o -path /opt/mattermost/plugins -o -path /opt/mattermost/data ')' -prune ')' -exec rm -r '{}' ';'
Unpacking mattermost (7.4.0-0) over (7.3.1-0) ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
Setting up mattermost (7.4.0-0) ...
+ IS_STRICT=false
+ [[ false =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ case "$1" in
+ '[' -z '' ']'
+ MATTERMOST_USER=mattermost
+ '[' -z '' ']'
+ MATTERMOST_GROUP=mattermost
+ getent group mattermost
+ id mattermost
+ chown -R mattermost:mattermost /opt/mattermost
+ '[' 7.3.1-0 '!=' '' ']'
+ IS_UPGRADE=true
+ [[ ! true =~ ^([tT][rR][uU][eE]|[yY]|[yY][eE][sS]|1)$ ]]
+ systemctl daemon-reload
+ systemctl start mattermost
```

Works great 🚀

spirosoik commented on mattermost/DEBIAN/postrm:
```diff
@@ -1,7 +1,19 @@
-        systemctl daemon-reload
-        ;;
+
+    [ -z "$MATTERMOST_USER" ] && MATTERMOST_USER="mattermost"
```
same as I mentioned above

phoinixgrr replied:
We are trying to perform the following:
If MATTERMOST_USER is null (not set) we set to to `mattermost`
Same for MATTERMOST_GROUP

spirosoik replied:
same as above

phoinixgrr replied:
Tested it, seems to work as expected.

What do we gain with the proposal?
Wouldn't you agree that it is a more readable form? :
[ -z "$MATTERMOST_USER" ] && MATTERMOST_USER="mattermost"
vs
MATTERMOST_USER=${MATTERMOST_USER:-mattermost}

The latter requires some bash insights.

Would vote in favor of readability against saving a couple of chars.

Let me know what do you think

spirosoik commented on mattermost/files/mattermost.service:
```diff
@@ -14,6 +12,7 @@ WorkingDirectory=/opt/mattermost
 User=mattermost
 Group=mattermost
 LimitNOFILE=49152
+Environment="MM_INSTALL_TYPE=ubuntu-testing"
```
is this the right env variable? the env value needs to change to the correct value right?

phoinixgrr replied:
Correct. Waiting for some feedback @ https://docs.google.com/document/d/1AXr0aip1pC1p62K-abDGX8rdWv9MtX9UiT8TLaz2iUY/edit?disco=AAAAh9_JbUg to finilize this. 

The functionality is implemented

spirosoik replied:
awesome, thanks for taking care of this 💪

spirosoik reviewed (changes requested):
few minor changes and recommendations. Nothing major

phoinixgrr commented on mattermost/files/mattermost.service:
```diff
@@ -1,8 +1,6 @@
 [Unit]
 Description=Mattermost
 After=network.target
-After=postgresql.service
```
We remove `postgresql` hardcoded dependency from mattermost-server packaging, since it is in the user's repsonsibility/choice to select his favorite db service (postgresql - mysql) 
Keeping it non opinionated. 

When shipping with omnibus-mattermost, we define this as postgresql (Highly opinionated)

spirosoik replied:
ok so for omnibus will stay as is got it.

phoinixgrr replied:
That's correct

spirosoik reviewed (approved):
LGTM 💪

phoinixgrr commented:
Applied the variable substitution changes requested 
```
MATTERMOST_USER=${MATTERMOST_USER:-mattermost}
MATTERMOST_GROUP=${MATTERMOST_GROUP:-mattermost}
```

Rebased and dropped your reviews @stafot  @spirosoik  
Thanks for re-reviewing

spirosoik reviewed (approved):
one more time awesome work @phoinixgrr
